polyfill/lib/ecmascript.mjs

```diff
@@ -3540,216 +3540,103 @@ export function UnbalanceDateDurationRelative(years, months, weeks, days, larges
+      if (years === 0) return { years: 0, months, weeks, days };
+      if (!calendarRec) throw new RangeError('a starting point is required for months balancing');
+      // balance years down to months
+      const later = CalendarDateAdd(calendarRec, plainRelativeTo, new TemporalDuration(years));
+      const untilOptions = ObjectCreate(null);
+      untilOptions.largestUnit = 'month';
+      const untilResult = CalendarDateUntil(calendarRec, plainRelativeTo, later, untilOptions);
+      const yearsInMonths = GetSlot(untilResult, MONTHS);
+      return { years: 0, months: months + yearsInMonths, weeks, days };
+    }
+    case 'week': {
+      if (years === 0 && months === 0) return { years: 0, months: 0, weeks, days };
+      if (!calendarRec) throw new RangeError('a starting point is required for weeks balancing');
+      // balance years and months down to days
+      const later = CalendarDateAdd(calendarRec, plainRelativeTo, new TemporalDuration(years, months));
+      const yearsMonthsInDays = DaysUntil(plainRelativeTo, later);
+      return { years: 0, months: 0, weeks, days: days + yearsMonthsInDays };
+    }
+    default: {
+      if (years === 0 && months === 0 && weeks === 0) return { years: 0, months: 0, weeks: 0, days };
+      if (!calendarRec) throw new RangeError('a starting point is required for balancing calendar units');
+      // balance years, months, and weeks down to days
+      const later = CalendarDateAdd(calendarRec, plainRelativeTo, new TemporalDuration(years, months, weeks));
+      const yearsMonthsWeeksInDays = DaysUntil(plainRelativeTo, later);
+      return { years: 0, months: 0, weeks: 0, days: days + yearsMonthsWeeksInDays };
+    }
   }
-
-  return {
-    years: years.toJSNumber(),
-    months: months.toJSNumber(),
-    weeks: weeks.toJSNumber(),
-    days: days.toJSNumber()
-  };
 }
 
 export function BalanceDateDurationRelative(years, months, weeks, days, largestUnit, plainRelativeTo, calendarRec) {
-  // dateAdd, dateUntil must be looked up if units are not 0 and largestUnit == year
-  // dateAdd must be looked up if units are not 0 and largestUnit == month or week
+  // calendarRec must have looked up:
+  //  - if largestUnit == year:
+  //    - if years or months nonzero, dateAdd and dateUntil
+  //    - if weeks or days nonzero, dateUntil
+  //  - if largestUnit == month:
+  //    - if months nonzero, dateAdd and dateUntil
+  //    - if weeks or days nonzero, dateUntil
+  //  - if largestUnit == week:
+  //    - if weeks nonzero, dateAdd and dateUntil
+  //    - if days nonzero, dateUntil
   const TemporalDuration = GetIntrinsic('%Temporal.Duration%');
-  const sign = DurationSign(years, months, weeks, days, 0, 0, 0, 0, 0, 0);
-  if (sign === 0 || (largestUnit !== 'year' && largestUnit !== 'month' && largestUnit !== 'week')) {
+
+  if (
+    (years === 0 && months === 0 && weeks === 0 && days === 0) ||
+    (largestUnit !== 'year' && largestUnit !== 'month' && largestUnit !== 'week')
+  ) {
```

Comment:
Feels like a comment could be useful to explain why an early return is used here.